### PR TITLE
Feature add variant_cast

### LINKED ISSUE #108: rttr::variant why only const getter?
why can we not get the object as non-const reference?

```cpp
rttr::variant instance = rttr::type::get_by_name("MyMessage").create();
const MyMessage& r = instance.get_value<MyMessage>();
```
I could use the policy as `rttr::policy::ctor::as_std_shared_ptr` or `rttr::policy::ctor::as_raw_ptr`.
then get the pointer (`MyMessage* & const`) . Is there a reason, why it returns a `const T&`?

Would it be possibe to have a policy which constructs a `std::unique_ptr`? Probably its not so easy as `std::move` is needed internally when we get the object. 

```
std::unique_ptr r =  instance.get_value<MyMessage>();
```
-> SFINAE which enables a `get_value` function which `std::move`s the stuff out?

Background:
I try to use this rttr stuff for serializing/deserializing messages, (serialize the type, here the `MyMessage` string, and the payload)

### PR DESCRIPTION
**Pull request by acki-m**
Added global "variant_cast" method

Added 4 overloads of the method "variant_cast"
In order to move values out, to return values by reference
or to return a pointer to the underlying values, when the type does match:

template<class T>
T variant_cast(variant&& operand);

template<class T>
T variant_cast(variant& operand);

template<class T>
T variant_cast(const variant& operand);

template<class T>
const T* variant_cast(const variant* operand) RTTR_NOEXCEPT;

template<class T>
T* variant_cast(variant* operand) RTTR_NOEXCEPT;

Additionally:
- added non-const version of variant::get_value()
- fixed copy-right year in root CMakeLists.txt file

Fixes #108

### CODE REVIEW

**Comment by acki-m** [comment]
@gabyx 
What do you think?

File: src/unit_tests/variant/variant_cast_test.cpp [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+        auto value = variant_cast<int>(&var);
+
+        CHECK(value == nullptr);
+    }
+
+    SECTION("valid cast")
+    {
+        variant var = 12;
+
+        auto value = variant_cast<int>(&var);
```
Comment (gabyx) [inline src/unit_tests/variant/variant_cast_test.cpp]: Is that correct: value is still bound to the `var`. meaning if var gets destroyed, `value` is dangling as well.

File: src/rttr/variant.h [hunk 2]
```diff
@@ -1067,13 +1098,107 @@ class RTTR_API variant
...
+
+/*!
+ * \brief Returns a pointer to the containing value with type \p T.
+ *        When the containing value is of type \p T, a valid pointer to the type will be returned.
+ *        Otherwise a `nullptr` is returned.
+ *
+ * \code{.cpp}
+ *
+ *  variant var = std::string("hello world");
+ *  std:string* a = variant_cast<std::string>(&var);  // performs an internal type check and returns extracts the value by reference
```
Comment (gabyx) [inline src/rttr/variant.h]: remove `extracts`

File: src/rttr/variant.h [hunk 3]
```diff
@@ -1067,13 +1098,107 @@ class RTTR_API variant
...
+ *  int* b        = variant_cast<int>(&var);
+ *  std::cout << "a valid: " << a != nullptr << std::endl;
+ *  std::cout << "b valid: " << b != nullptr << std::endl;
```
Comment (gabyx) [inline src/rttr/variant.h]: Probably add "// b is nullptr"

File: src/unit_tests/variant/variant_cast_test.cpp [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+
+        CHECK(value == nullptr);
+    }
+
+    SECTION("valid cast")
+    {
+        const variant var = 12;
+
+        auto value = variant_cast<int>(&var);
```
Comment (gabyx) [inline src/unit_tests/variant/variant_cast_test.cpp]: so the const semantic  in `variant_cast<...>` is neglected. meaning we dont write `variant_cast<const int>`.
Wouldnt that be better? I see the benefits of the current implementation. I would leave it like that

File: src/unit_tests/variant/variant_cast_test.cpp [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+
+TEST_CASE("variant - variant_cast<T>(variant&)", "[variant]")
+{
+    SECTION("by value")
+    {
+        variant var = 12;
+
+        REQUIRE(var.get_type() == type::get<int>());
+
+        auto value = variant_cast<int>(var);
```
Comment (gabyx) [inline src/unit_tests/variant/variant_cast_test.cpp]: ;) -> `auto` has the semantic of always stripping `references` == `&`. So I think when you write `auto value` this is always an lvalue (not an lvalue reference). 
So it seems that this test is obsolete :-), but better leave it included :-)

File: src/rttr/variant.h [hunk 6]
```diff
@@ -423,6 +428,32 @@ class RTTR_API variant
...
+         *         Use therefore the method \ref is_type().
+         *         Otherwise the call leads to undefined behaviour.
+         *         Also make sure you don't clean this variant, when you still hold a reference to the containing value.
+         *
+         * \see is_type()
+         *
+         * \return A reference to the stored value.
+         */
+        template<typename T>
+        T& get_value();
```
Comment (gabyx) [inline src/rttr/variant.h]: I like that one. So such a functionality was not yet there. Its up to the user to call that with the right type.

File: src/rttr/detail/variant/variant_data_policy.h [hunk 7]
```diff
@@ -652,7 +652,7 @@ struct RTTR_API variant_data_policy_empty
             }
             case variant_policy_operation::GET_VALUE:
             {
-                arg.get_value<void*>() = nullptr;
+                arg.get_value<const void*>() = nullptr;
```
Comment (gabyx) [inline src/rttr/detail/variant/variant_data_policy.h]: was that wrong before? and did you already had a get_value function before but not in the interface in `variant` (I suppose `arg` is a variant?)

File: src/unit_tests/variant/variant_cast_test.cpp (hunk 1 above)
Comment (acki-m) [reply src/unit_tests/variant/variant_cast_test.cpp]: yes, that is correct. Same as with `variant::get_value<T>()`

File: src/rttr/detail/variant/variant_data_policy.h (hunk 7 above)
Comment (acki-m) [reply src/rttr/detail/variant/variant_data_policy.h]: I go down with a `const void*`, that's why I adjusted it.

File: src/unit_tests/variant/variant_cast_test.cpp (hunk 5 above)
Comment (acki-m) [reply src/unit_tests/variant/variant_cast_test.cpp]: I know, but also this case can happen, thats why I added the test.
Its also in the section: `SECTION("by value")`

File: src/unit_tests/variant/variant_cast_test.cpp (hunk 4 above)
Comment (acki-m) [reply src/unit_tests/variant/variant_cast_test.cpp]: @gabyx 
What would be better?
You cannot have a `const-by-value` inside a variant. However, you can have a `const int*`

**Pull request merged**